HP-137: 메인 페이지 및 상품 관련 주요 컴포넌트 추가 및 로그인 상태 관리 개선

minseung-gang opened the pull request:
## PR 타입

- [x] 기능 추가
- [ ] 버그 수정
- [ ] 리팩토링
- [ ] 문서 수정
- [ ] 기타

## 반영 브랜치

HP-137 → dev

## 작업 사항
- 메인 페이지 무한 스크롤 및 유저 초기 진입 모달 처리 추가
- 로그인 상태 동기화를 위한 useAuthStateManager 훅 구현
- 유저 정보 조회 및 닉네임 설정을 위한 React Query 훅 구현
- queryKey에 productList 및 user 관련 키 추가
- 상품 리스트 및 베스트 상품 조회를 위한 React Query 훅 추가
- CategoryButton에 Skeleton UI 컴포넌트 추가
- 상품 정보를 보여주는 Card 컴포넌트 및 Skeleton UI 추가
- 장바구니 아이콘 버튼 CartButton 컴포넌트 추가
- 카테고리 선택용 CategoryButton 컴포넌트 구현
- 메인 페이지에 Section 레이아웃 컴포넌트 추가
- TagButton 컴포넌트 추가
- CategorySection 컴포넌트 추가
- 메인 페이지 MainBanner, PromotionBanner 컴포넌트 구현
- ProductList 및 ProductSection 컴포넌트 추가 및 개선
- useLoginedStore 상태 타입 boolean으로 변경 및 setLogined 함수명 수정

## 체크리스트

- [x] 빌드에 성공했나요?
- [x] 코드 컨벤션을 잘 지켰나요? (예: `cmd + opt + L`)

## 테스트 결과
- 테스트 결과 이상 없습니다.

minseung-gang committed:
feat: 메인 페이지 무한 스크롤 및 유저 초기 진입 모달 처리 추가
- 닉네임이 없는 유저일 경우 'welcomeStep' 모달 자동 오픈 처리
- 카테고리 변경 시 활성 카테고리 ID 업데이트 로직 추가
- 상품 리스트 무한 스크롤 시 최대 5회까지 재시도 로직 추가
- 마지막 페이지가 비어 있는 경우 자동으로 다음 페이지 요청하도록 useEffect 처리

minseung-gang committed:
feat: 로그인 상태 동기화를 위한 useAuthStateManager 훅 구현

minseung-gang committed:
feat: 유저 정보 조회 및 닉네임 설정을 위한 React Query 훅 구현

minseung-gang committed:
feat: queryKey에 productList 및 user 관련 키 추가

minseung-gang committed:
feat: 상품 리스트 및 베스트 상품 조회를 위한 React Query 훅 추가

minseung-gang committed:
feat: CategoryButton에 Skeleton UI 컴포넌트 추가

minseung-gang committed:
feat: 상품 정보를 보여주는 Card 컴포넌트 추가

minseung-gang committed:
feat: Card 컴포넌트용 Skeleton UI 추가

minseung-gang committed:
feat: 장바구니 아이콘 버튼 CartButton 컴포넌트 추가

minseung-gang committed:
feat: 카테고리 선택용 CategoryButton 컴포넌트 구현

minseung-gang committed:
feat: 메인 페이지에 Section 레이아웃 컴포넌트 추가

minseung-gang committed:
feat: TagButton 컴포넌트 추가

minseung-gang committed:
feat: CategorySection 컴포넌트 추가
 - 카테고리 버튼 리스트 및 로딩 스켈레톤 처리

minseung-gang committed:
feat: 메인 페이지 MainBanner 컴포넌트 구현

minseung-gang committed:
feat: ProductList 컴포넌트 추가
- 상품 목록, 로딩 및 빈 상태 UI 구현

minseung-gang committed:
feat: PromotionBanner 컴포넌트 추가
- 프로모션 배너 캐러셀 구현

minseung-gang committed:
refactor: useLoginedStore isLogined 타입 boolean으로 변경 및 setLogined 함수명 수정

minseung-gang committed:
feat: ProductSection 컴포넌트 추가
- 카테고리 및 상품 리스트 통합 UI 구현

minseung-gang committed:
refactor: useAuthStateManager 로그인 상태 업데이트 로직 개선

minseung-gang committed:
fix: 닉네임 존재 여부에 따른 로그인 상태 로직 수정

minseung-gang committed:
feat: PromotionBanner에 사용 될 이미지 추가

minseung-gang committed:
feat: BestProductSection 컴포넌트 구현 및 로딩, 빈 상태 처리 추가

minseung-gang committed:
refactor: 자동 페이징 관련 useEffect 제거 (버튼 기반 로딩 구조에 불필요한 코드 삭제)

minseung-gang committed:
feat: 닉네임 검증을 포함한 3단계 인증 상태 관리 구현
- userData 없을 때: setLogined(null)
- userData 있지만 nickname 없을 때: 메인페이지 이동
- userData와 nickname 모두 있을 때: setLogined(true)

minseung-gang committed:
refactor: 메인페이지 경로 예외처리

minseung-gang committed:
feat: 카테고리명 해시 기반 배경색 매핑 함수 구현 (임시)

minseung-gang committed:
Merge branch 'dev' into HP-137/ms

jieun419 commented on src/pages/main/components/atoms/Card.tsx:
```diff
@@ -0,0 +1,50 @@
+          {best && <TagButton value='BEST' bgColor={bestTagBgColor} />}
+          <TagButton value={categoryName} bgColor={categoryTagBgColor} />
+        </div>
+        <a href={`/product/${product.productId}`}>
```
react-router-dom에서 제공하는 `Link`태그를 통해 페이지 이동하는게 맞을 것 같습니다!

jieun419 commented on src/pages/main/components/atoms/CartButton.tsx:
```diff
@@ -0,0 +1,18 @@
+
+import Button from '@/components/button/BaseButton';
+
+type CartButtonProps = ButtonHTMLAttributes<HTMLButtonElement>;
```
props 등의 타입 지정 시 `interface`로 컨벤션 통일하면 좋을 것 같습니다!

jieun419 commented on src/pages/main/components/atoms/CategoryButton.tsx:
```diff
@@ -0,0 +1,40 @@
+
+    return (
+      <Button
+        className={cn(buttonClasses, className)}
```
buttonClasses에 이미 `cn`을 선언한 상태로 `className`을 `buttonClasses` 내부 데이터에 같이 두는건 어떠신가요?

```tsx
const buttonClasses = cn(
      'w-full min-w-fit flex justify-center items-center transition-colors duration-100',
      {
        'bg-primary-text text-white': isActive,
        'bg-white text-primary-text': !isActive,
      },
      className
    );
.
.
.

<Button className={buttonClasses}
```

jieun419 commented on src/pages/main/components/atoms/CategoryButton.tsx:
```diff
@@ -0,0 +1,40 @@
+  icon?: string;
+  isActive: boolean;
+  className?: string;
+  onClickHandler: () => void;
```
다른 컴포넌트 처럼 onClickHandler 대신 `onClick`으로 변경하면 좋을 것 같아요!

jieun419 commented on src/pages/main/components/molecules/Category.tsx:
```diff
@@ -0,0 +1,68 @@
+      <div className='hide-scrollbar md:gap-x-auto mb-5 flex w-full gap-x-3 gap-y-4 overflow-x-scroll px-4 md:grid md:w-48 lg:px-0'>
+        {renderContent()}
+      </div>
+    </div>
```
의미 없는 `div` 태그가 삽입 되어 있습니다.
삭제해도 될 것 같습니다!

jieun419 commented on src/pages/main/components/molecules/MainBanner.tsx:
```diff
@@ -0,0 +1,57 @@
+        variant='centered'
+        centerPadding={window.innerWidth < 768 ? 30 : 0}
+        isInfinite
+        gap={window.innerWidth < 768 ? 15 : 20}
```
768라는 매직넘버를 상수로 관리하면 가독성이 좋을 것 같습니다!

ex- MIN_WIDTH = 768

jieun419 commented on src/pages/main/components/molecules/PromotionBanner.tsx:
```diff
@@ -0,0 +1,49 @@
+    <div className='my-25 bg-[#BED0A2] lg:my-36'>
+      <Carousel
+        variant='centered'
+        centerPadding={window.innerWidth < 768 ? 10 : 0}
```
아까 코멘트 처럼 768 매직 넘버 상수로 관리하면 가독성 + 유지보수에 좋을 것 같습니다!

jieun419 commented on src/pages/main/components/organisms/BestProduct.tsx:
```diff
@@ -0,0 +1,64 @@
+
+  return (
+    <Section title='best selection' className='mt-25 px-4 md:mt-20 lg:mt-48'>
+      <div className='mx-auto w-full max-w-[1280px]'>{renderContent()}</div>
```
`max-w-[1280px]`는 index.css에 선언되어 있는 공통 클래스를 사용하면 좋을 것 같습니다!

jieun419 commented on src/hooks/api/member/product.ts:
```diff
@@ -1,19 +1,101 @@
+      };
+    },
+
+    initialPageParam: { categoryIndex: 0, lastProductId: undefined } as PageParam,
```
`lastProductId` 는 초기값으로  undefined대신 `null`로 명시해도 좋을 것 같습니다!

jieun419 commented on src/pages/main/components/atoms/CardSkeleton.tsx:
- 반복되는 ui는 Arroy.form을 활용해 맵핑하면 좋을 것 같습니다.
- div 내부에 아무런 내용이 없다면 `<div />` 형식으로 닫기 태그를 없애도 될 것 같아요!

jieun419 commented on src/pages/main/components/organisms/Product.tsx:
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
```
React선언하는 곳이 없다면 
`import React from 'react';` 삭제하셔도 될 것 같습니다!

minseung-gang replied:
넵! 지은님 React는 컴포넌트 타입을 선언할 때 사용했습니다!

jieun419 reviewed (commented):
고생하셨습니다!

- 스켈레톤 ui는 추후 공통 컴포넌트 또는 클래스를 만들어 해당 요소를 가져와 활용하면 좋을 것 같습니다!
- 코멘트에서 매직 넘버를 상수로 관리하는 내용은 constants에 파일 생성의미가 아닌 해당 컴포넌트 내부에서 관리하는걸 의미합니다! (오해하실 까봐..)

minseung-gang commented:
지은님 코멘트 잘 봤습니다! 리뷰 감사드려요 ㅎㅎ
스켈레톤에 관련해서는 추후 이 컴포넌트가 다른 화면에서도 반복된다면, 공통 컴포넌트로 분리하는 것도 고려해보면 좋을 것 같습니다 :)

minseung-gang committed:
refactor: a태그를 Link 컴포넌트로 변경
- SPA 네비게이션으로 페이지 새로고침 방지
- React Router의 히스토리 관리 활용

minseung-gang committed:
refactor: CartButton props에 ButtonHTMLAttributes 타입 명시로 타입 안정성 강화 및 컨벤션 준수

minseung-gang committed:
refactor: CategoryButton onClickHandler prop명을 onClick으로 변경해 통일성 유지
- buttonClasses에 className을 cn 함수 내부 인자로 포함하여 코드 간결화
- 이벤트 핸들러 prop명 통일로 컴포넌트 사용 일관성 향상

minseung-gang committed:
refactor: CategorySection onClick prop명 변경 및 불필요한 div 제거

minseung-gang committed:
refactor: 768px 브레이크포인트를 상수로 관리하여 가독성과 유지보수성 향상

minseung-gang committed:
refactor: div에 적용된 max-w 스타일을 커스텀 클래스명으로 변경하여 스타일 관리 일관성 개선

minseung-gang committed:
refactor: lastProductId prop 타입에 null 추가하여 초기값 명확화

minseung-gang committed:
refactor: CardSkeleton 컴포넌트 리팩토링
- 반복되는 div 태그를 배열과 map으로 간결하게 처리
- 내용 없는 div 태그를 self-closing 태그로 변경하여 코드 가독성 향상

minseung-gang committed:
Merge branch 'HP-137/ms' of https://github.com/happy-pill/happyPill_Front into HP-137/ms

minseung-gang committed:
feat: 유저 API 추가 및 닉네임 수정 기능 구현

jieun419 reviewed (approved):
고생하셨슴다!
